File: 백준/Silver/2805. 나무 자르기/나무 자르기.cc
#include <algorithm>
#include <iostream>

using namespace std;

bool is_possible(int *tree, int N, int M, int height) {
  long long sum = 0;
  for (int i = 0; i < N; i++) {
    if (tree[i] > height) {
      sum += tree[i] - height;
    }
  }
  return sum >= M;
}

int main() {
  int N, M;
  cin >> N >> M;

  int *tree = new int[N];

  for (int i = 0; i < N; i++) {
    cin >> tree[i];
  }
  sort(tree, tree + N);

  int low = 0;
  int high = tree[N - 1];

  while (low + 1 < high) {
    int mid = (low + high) / 2;
    if (is_possible(tree, N, M, mid)) {
      low = mid;
    } else {
      high = mid;
    }
  }

  cout << low << endl;
}

File: 백준/Silver/2805. 나무 자르기/나무 자르기.py
import sys

input = sys.stdin.readline

N, M = map(int, input().split())
trees = list(map(int, input().split()))
h = max(trees) // 2

def is_possible(len):
  return M <= sum(map(lambda x: 0 if x <= len else x - len, trees))

low = 0
high = max(trees) - 1
while low + 1 < high:
  mid = (low + high) // 2
  if is_possible(mid):
    low = mid
  else:
    high = mid

if is_possible(high):
  print(high)
else:
  print(low)